pylint/testutils/configuration_test.py

```diff
@@ -0,0 +1,101 @@
+EXPECTED_CONF_REMOVE_KEY = "functional_remove"
+EXPECTED_CONF_SPECIAL_KEYS = [EXPECTED_CONF_APPEND_KEY, EXPECTED_CONF_REMOVE_KEY]
+
+
+def get_expected_configuration(
+    configuration_path: str, default_configuration: Dict[str, Any]
+) -> Dict[str, Any]:
+    """Get the expected parsed configuration of a configuration functional test"""
+    result = copy.deepcopy(default_configuration)
+    config_as_json = get_expected_or_default(
+        configuration_path, suffix="result.json", default="{}"
+    )
+    to_override = json.loads(config_as_json)
+    for key, value in to_override.items():
+        if key not in EXPECTED_CONF_SPECIAL_KEYS:
+            result[key] = value
+    if EXPECTED_CONF_APPEND_KEY in to_override:
+        for key, value in to_override[EXPECTED_CONF_APPEND_KEY].items():
+            result[key] += value
+    if EXPECTED_CONF_REMOVE_KEY in to_override:
+        for key, value in to_override[EXPECTED_CONF_REMOVE_KEY].items():
+            new_value = []
+            for old_value in result[key]:
+                if old_value not in value:
+                    new_value.append(old_value)
+            result[key] = new_value
+    return result
+
+
+def get_expected_output(configuration_path: str) -> Tuple[int, str]:
+    """Get the expected output of a functional test."""
+
+    def get_relative_path(path: str) -> str:
+        """Get the relative path we want without user specific information"""
+        return path.replace(USER_SPECIFIC, "")[1:]
+
+    output = get_expected_or_default(configuration_path, suffix="out", default="")
+    if output:
+        logging.info(
+            "Output exists for %s so the expected exit code is 2", configuration_path
+        )
+        exit_code = 2
+    else:
+        logging.info(".out file does not exists, so the expected exit code is 0")
+        exit_code = 0
+    return exit_code, output.format(
+        abspath=configuration_path, relpath=get_relative_path(configuration_path)
+    )
+
+
+def run_using_a_configuration_file(
+    configuration_path: Union[Path, str], file_to_lint: str = __file__
+) -> Tuple[Mock, Mock, Run]:
+    """Simulate a run with a configuration without really launching the checks."""
+    configuration_path = str(configuration_path)
+    args = ["--rcfile", configuration_path, file_to_lint]
+    # If we used `pytest.raises(SystemExit)`, the `runner` variable
+    # would not be accessible outside the `with` block.
+    with unittest.mock.patch("sys.exit") as mocked_exit:
+        # Do not actually run checks, that could be slow. Do not mock
+        # `Pylinter.check`: it calls `Pylinter.initialize` which is
+        # needed to properly set up messages inclusion/exclusion
+        # in `_msg_states`, used by `is_message_enabled`.
+        check = "pylint.lint.pylinter.check_parallel"
+        with unittest.mock.patch(check) as mocked_check_parallel:
+            runner = Run(args)
+    return mocked_exit, mocked_check_parallel, runner
```

Comment:
We never seem to use `mocked_check_parallel`. Do we need to return it?

Reply:
We could remove it, but this might permit to do additional check as it's a test util function that could be used by lib depending on pylint for their own tests, so I thought we might as well provide it.

Reply:
Just wanted to check this was not left from an earlier version in which you did use the second return value 😄

Reply:
In fact it's taken from [here](https://github.com/PyCQA/pylint/blob/main/tests/config/test_config.py#L15), I'll refactor and move everything to functional in a subsequent MR. Or maybe I should do it right now.

Reply:
https://github.com/PyCQA/pylint/pull/5287/commits/7814c2fbb47a8a9f2824af77a7154201cd240964